test/integration/apiserver.go

```diff
@@ -0,0 +1,359 @@
+	server := apiserver.NewAPI(apiserver.APIServiceOptions{
+		Version:  kube.IstioResourceVersion,
+		Port:     8081,
+		Registry: &model.IstioRegistry{ConfigRegistry: controller},
+	})
+	r.stopChannel = make(chan struct{})
+	go controller.Run(r.stopChannel)
+	go server.Run()
+
+	// Wait until apiserver is ready.  (As far as I can see there is no ready channel)
+	for i := 0; i < 10; i++ {
+		_, err := net_http.Get("http://localhost:8081/")
+		if err == nil {
+			break
+		}
+		time.Sleep(1 * time.Second)
+	}
+
+	testURL = "http://localhost:8081/v1alpha1/config/route-rule/" + r.Namespace + "/reviews-default"
+
+	return nil
+}
+
+func (r *apiServerTest) teardown() {
+	if r.stopChannel != nil {
+		close(r.stopChannel)
+	}
+}
+
+func (r *apiServerTest) run() error {
+	if err := r.routeRuleInvalidDetected(); err != nil {
+		return err
+	}
+	if err := r.routeRuleCRUD(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *apiServerTest) routeRuleInvalidDetected() error {
+	httpSequence := []httpRequest{
+		// Can't create invalid rules
+		{
+			method: "POST", url: "http://localhost:8081/v1alpha1/config/route-rule/" + r.Namespace + "/reviews-invalid",
+			data:                 jsonInvalidRule,
+			expectedResponseCode: net_http.StatusInternalServerError,
+		},
+	}
+
+	return verifySequence(httpSequence)
+}
+
+// routeRuleCRUD attempts to talk to the apiserver to Create, Retrieve, Update, and Delete a rule
+func (r *apiServerTest) routeRuleCRUD() error {
+
+	// Run through the lifecycle of updating a rule, with errors, to verify the sequence works as expected.
+	httpSequence := []httpRequest{
+		// Step 0 Can't get before created
+		{
+			method: "GET", url: testURL,
+			expectedResponseCode: net_http.StatusNotFound,
+		},
+		// Step 1 Can create
+		{
+			method: "POST", url: testURL,
+			data:                 jsonRule,
+			expectedResponseCode: net_http.StatusCreated,
+			expectedBody:         jsonRule,
+			// @@ retryOn:              map[int]bool{net_http.StatusInternalServerError: true},
+		},
+		// Step 2 Can't create twice
+		{
+			// TODO should be StatusForbidden but the server is returning 500 (incorrectly?)
+			method: "POST", url: testURL,
+			data:                 jsonRule,
+			expectedResponseCode: net_http.StatusInternalServerError,
+		},
+		// Step 3 Can get
+		{
+			method: "GET", url: testURL,
+			expectedResponseCode: net_http.StatusOK,
+			expectedBody:         jsonRule,
+			retryOn:              map[int]bool{net_http.StatusNotFound: true},
+		},
+		// Step 4: Can update
+		{
+			method: "PUT", url: testURL,
+			data:                 jsonRule2,
+			expectedResponseCode: net_http.StatusOK,
+			expectedBody:         jsonRule2,
+		},
+		// Can still GET after update
+		{
+			method: "GET", url: testURL,
+			expectedResponseCode: net_http.StatusOK,
+			expectedBody:         jsonRule2,
+			retryOn:              map[int]bool{net_http.StatusNotFound: true},
+			retryIfMatch:         []interface{}{jsonRule},
+		},
```

Comment:
Add a second config then do a GET `config/route-rules/namespace` as well because it's a separate codepath